### PR TITLE
[AMDGPU] Add amdgpu-as MMRA for fences

### PR DESCRIPTION
**Pull request by Pierre-vh**
Using MMRAs, allow `builtin_amdgcn_fence` to emit fences that only target one or more address spaces, instead of fencing all address spaces at once.

This is done through a `amdgpu-as` MMRA. Currently focused on OpenCL fences, but can very easily support more AS names and codegen on more than just fences.

### CODE REVIEW

File: clang/include/clang/Basic/BuiltinsAMDGPU.def [hunk 1]
```diff
@@ -68,6 +68,7 @@ BUILTIN(__builtin_amdgcn_iglp_opt, "vIi", "n")
 BUILTIN(__builtin_amdgcn_s_dcache_inv, "v", "n")
 BUILTIN(__builtin_amdgcn_buffer_wbinvl1, "v", "n")
 BUILTIN(__builtin_amdgcn_fence, "vUicC*", "n")
+BUILTIN(__builtin_amdgcn_fence_opencl, "vUiUicC*", "n")
```
Comment (arsenm) [inline clang/include/clang/Basic/BuiltinsAMDGPU.def:71]: We shouldn't have language specific target builtins. Why isn't the existing fence adequate?

File: clang/lib/CodeGen/CGBuiltin.cpp [hunk 2]
```diff
@@ -18319,6 +18320,26 @@ Value *CodeGenFunction::EmitHLSLBuiltinExpr(unsigned BuiltinID,
...
+  constexpr const char *Tag = "opencl-fence-mem";
+
+  uint64_t Mask = cast<llvm::ConstantInt>(ASMask)->getZExtValue();
+  if (Mask == 0)
+    return;
+
+  // 3 bits can be set: local, global, image in that order.
+  LLVMContext &Ctx = Inst->getContext();
+  SmallVector<MMRAMetadata::TagT, 3> MMRAs;
+  if (Mask & (1 << 0))
```
Comment (ssahasra) [inline clang/lib/CodeGen/CGBuiltin.cpp:18334]: These magic numbers for the shift amount should be declared as an Enum somewhere.

File: clang/include/clang/Basic/BuiltinsAMDGPU.def (hunk 1 above)
Comment (ssahasra) [reply clang/include/clang/Basic/BuiltinsAMDGPU.def:71]: I think the motivation here is that we now have an integer "address space mask" argument to the fence. The default should be `0`, to represent "all address spaces", but it can be a smaller subset. This is argument gets translated to the MMRAs. But this shouldn't be OpenCL-specific. We need a scheme that works with all input languages that we care about.

File: clang/lib/CodeGen/CGBuiltin.cpp (hunk 2 above)
Comment (Pierre-vh) [reply clang/lib/CodeGen/CGBuiltin.cpp:18369]: Agreed, but not sure where they should go to be meaningful. Any ideas?
Maybe `TargetParser` could have a list of MMRA-related values (tag prefix/suffixes, mask values for builtins, etc.) ?

Comment (arsenm) [reply clang/lib/CodeGen/CGBuiltin.cpp:18369]: Alternatively could behave more like system scopes where the builtin takes a variadic list of string literal arguments passed directly through

File: clang/include/clang/Basic/BuiltinsAMDGPU.def [hunk 3]
```diff
@@ -69,6 +69,7 @@ BUILTIN(__builtin_amdgcn_iglp_opt, "vIi", "n")
 BUILTIN(__builtin_amdgcn_s_dcache_inv, "v", "n")
 BUILTIN(__builtin_amdgcn_buffer_wbinvl1, "v", "n")
 BUILTIN(__builtin_amdgcn_fence, "vUicC*", "n")
+BUILTIN(__builtin_amdgcn_masked_fence, "vUiUicC*", "n")
```
Comment (arsenm) [inline clang/include/clang/Basic/BuiltinsAMDGPU.def:72]: Should put some documentation in LanguageExtensions for the arguments

File: clang/lib/CodeGen/CGBuiltin.cpp (hunk 2 above)
Comment (Pierre-vh) [reply clang/lib/CodeGen/CGBuiltin.cpp:18369]: @arsenm Do you mean just having a space-separate list of address spaces such as `"image local"`?
If that's acceptable then it definitely works for me too

File: clang/include/clang/Basic/BuiltinsAMDGPU.def (hunk 3 above)
Comment (Pierre-vh) [reply clang/include/clang/Basic/BuiltinsAMDGPU.def:72]: Added docs with tests. I also documented the other fence intrinsic

File: clang/docs/LanguageExtensions.rst [hunk 4]
```diff
@@ -4403,6 +4404,60 @@ Target-Specific Extensions
...
+* ``unsigned`` address space mask
+* ``unsigned`` atomic ordering, e.g. ``__ATOMIC_ACQUIRE``
+* ``const char *`` synchronization scope, e.g. ``workgroup``
+
+The address space mask is a bitmask and each bit corresponds
+to an OpenCL memory type:
+
+* ``0x1`` (first bit) is for local memory.
+* ``0x2`` (second bit) is for global memory.
+* ``0x4`` (third bit) is for image memory.
```
Comment (arsenm) [inline clang/docs/LanguageExtensions.rst:4439]: Should either get named enums or string-like treatment like the sync scope

**Comment by arsenm** [comment]
> I'm now wondering if adding a new builtin is needed at all, or if it should just be part of the original builtin? It's an additive change.

Maybe?

> 
> Should we also rename the MMRA to `amdgpu-fence-as` (remove OpenCL from the name) ?
> 

I definitely do not want to maintain any language names in anything

File: clang/docs/LanguageExtensions.rst [hunk 5]
```diff
@@ -4408,6 +4409,54 @@ Target-Specific Extensions
...
+* ``const char *`` synchronization scope, e.g. ``workgroup``
+* Zero or more ``const char *`` address spaces.
+
+The address spaces arguments must be string literals with known values, such as:
+
+* ``"local"``
+* ``"global"``
+* ``"image"``
+
+If there are no address spaces specified, this fence behaves like
```
Comment (ssahasra) [inline clang/docs/LanguageExtensions.rst:4445]: To answer comments elsewhere, this documentation makes it really obvious that the two fences are really just one.

File: clang/lib/CodeGen/CGBuiltin.cpp [hunk 6]
```diff
@@ -18365,6 +18366,30 @@ Value *CodeGenFunction::EmitHLSLBuiltinExpr(unsigned BuiltinID,
...
+                                                unsigned FirstASNameIdx) {
+  constexpr const char *Tag = "opencl-fence-mem";
+
+  LLVMContext &Ctx = Inst->getContext();
+  SmallVector<MMRAMetadata::TagT, 3> MMRAs;
+  for (unsigned K = FirstASNameIdx; K < E->getNumArgs(); ++K) {
+    llvm::Value *V = EmitScalarExpr(E->getArg(K));
+    StringRef AS;
+    if (llvm::getConstantStringInfo(V, AS) &&
+        (AS == "local" || AS == "global" || AS == "image")) {
```
Comment (ssahasra) [inline clang/lib/CodeGen/CGBuiltin.cpp:18380]: Can these magic strings be declared somewhere? Ideally some sort of target-specific mechanism that Clang will use to find such sets of valid strings?

File: clang/lib/CodeGen/CGBuiltin.cpp [hunk 7]
```diff
@@ -18365,6 +18366,30 @@ Value *CodeGenFunction::EmitHLSLBuiltinExpr(unsigned BuiltinID,
   return nullptr;
 }
 
+void CodeGenFunction::AddAMDGCNAddressSpaceMMRA(llvm::Instruction *Inst,
+                                                const CallExpr *E,
+                                                unsigned FirstASNameIdx) {
+  constexpr const char *Tag = "opencl-fence-mem";
```
Comment (ssahasra) [inline clang/lib/CodeGen/CGBuiltin.cpp:18372]: Defintely drop the "opencl" prefix.

**Comment by ssahasra** [comment]
> Should we also rename the MMRA to `amdgpu-fence-as` (remove OpenCL from the name) ?

Even the "fence" prefix is not entirely correct. The same tags also make sense on a load-acquire or store-release, which are "fence like" instructions, or "operations with implicit fences". Why not just "as:global", "as:local", etc? The fact that they are used as !mmra on a fence-like instruction makes it clear that they represent the address spaces that are caused to be synchronized by that instruction, and not incidentally the address space that the load-acquire or store-release itself wants to access.

File: clang/docs/LanguageExtensions.rst [hunk 8]
```diff
@@ -4408,6 +4409,42 @@ Target-Specific Extensions
...
+* Zero or more ``const char *`` address spaces names.
+
+The address spaces arguments must be string literals with known values, such as:
+
+* ``"local"``
+* ``"global"``
+* ``"image"``
+
+If one or more address space name are provided, the code generator will attempt
+to emit potentially faster instructions that only fence those address spaces.
```
Comment (ssahasra) [inline clang/docs/LanguageExtensions.rst:4431]: This use of "fence" as a verb seems a bit too informal. Reword it to say "instructions that order access to at least those address spaces"? (Note the addition of "at least" to signify a lower bound)

File: clang/lib/CodeGen/CGBuiltin.cpp [hunk 9]
```diff
@@ -18365,6 +18366,28 @@ Value *CodeGenFunction::EmitHLSLBuiltinExpr(unsigned BuiltinID,
   return nullptr;
 }
 
+void CodeGenFunction::AddAMDGCNFenceAddressSpaceMMRA(llvm::Instruction *Inst,
```
Comment (ssahasra) [inline clang/lib/CodeGen/CGBuiltin.cpp:18369]: The function immediately below this uses "AMDGPU" in its name. I think that's the newer practice. I don't have a strong opinion on this, because the same file also has functions which say "AMDGCN" instead. I am not sure which way the naming convention is leaning.

File: llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp [hunk 10]
```diff
@@ -678,6 +679,59 @@ class SIMemoryLegalizer final : public MachineFunctionPass {
...
+    {"image", SIAtomicAddrSpace::SCRATCH},
+}};
+
+void diagnoseUnknownMMRAASName(const MachineInstr &MI, StringRef AS) {
+  const MachineFunction *MF = MI.getMF();
+  const Function &Fn = MF->getFunction();
+  std::string Str;
+  raw_string_ostream OS(Str);
+  OS << "unknown address space '" << AS << "'; expected one of ";
+  bool IsFirst = true;
```
Comment (ssahasra) [inline llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp:694]: Use ListSeparator from StringExtras.h

**Review by ssahasra — APPROVED** [review]
The frontend changes and the MMRA emitted in LLVM IR look good to me. The backend changes also look okay, but please see if anyone else has comments about that.

File: clang/lib/CodeGen/CGBuiltin.cpp [hunk 11]
```diff
@@ -18365,6 +18366,28 @@ Value *CodeGenFunction::EmitHLSLBuiltinExpr(unsigned BuiltinID,
...
+                                                     const CallExpr *E) {
+  constexpr const char *Tag = "amdgpu-as";
```
Comment (ssahasra) [inline clang/lib/CodeGen/CGBuiltin.cpp:18371]: Just bikeshedding a bit, but do we really need the "amdgpu" prefix on the tag? Clang will only generate these for AMDGPU anyway. It's not a blocker, but feels like we are being cautious for no reason.

File: clang/docs/LanguageExtensions.rst [hunk 12]
```diff
@@ -4408,6 +4409,42 @@ Target-Specific Extensions
...
+If no address spaces names are provided, all address spaces are fenced.
+
+.. code-block:: c++
+
+  // Fence all address spaces.
+  __builtin_amdgcn_fence(__ATOMIC_SEQ_CST, "workgroup");
+  __builtin_amdgcn_fence(__ATOMIC_ACQUIRE, "agent");
+
+  // Fence only requested address spaces.
+  __builtin_amdgcn_fence(__ATOMIC_SEQ_CST, "workgroup", "local")
```
Comment (arsenm) [inline clang/docs/LanguageExtensions.rst:4444]: We randomly change between HSA and OpenCL terminology. Maybe we should call "local" "groupsegment"? I guess the ISA manuals call it "local data share"

File: clang/test/CodeGenCXX/builtin-amdgcn-fence.cpp [hunk 13]
```diff
@@ -1,22 +1,113 @@
...
+// CHECK-SAME: ) #[[ATTR0]] {
+// CHECK-NEXT:  entry:
+// CHECK-NEXT:    fence syncscope("workgroup") seq_cst, !mmra [[META6:![0-9]+]]
+// CHECK-NEXT:    fence syncscope("workgroup") seq_cst, !mmra [[META7:![0-9]+]]
+// CHECK-NEXT:    ret void
+//
+void test_mixed() {
+  __builtin_amdgcn_fence( __ATOMIC_SEQ_CST, "workgroup", "image", "global");
+  __builtin_amdgcn_fence( __ATOMIC_SEQ_CST, "workgroup", "image", "local", "global");
+}
```
Comment (arsenm) [inline clang/test/CodeGenCXX/builtin-amdgcn-fence.cpp:106]: Maybe test repeated AS name

File: clang/lib/CodeGen/CGBuiltin.cpp (hunk 11 above)
Comment (Pierre-vh) [reply clang/lib/CodeGen/CGBuiltin.cpp:18371]: I think it's just more readable this way, but I have no strong preference either.
I'd say let's keep it this way, it can very easily be renamed later if we want to.

File: clang/docs/LanguageExtensions.rst (hunk 12 above)
Comment (Pierre-vh) [reply clang/docs/LanguageExtensions.rst:4444]: I wasusing the OpenCL terminology: https://registry.khronos.org/OpenCL/sdk/3.0/docs/man/html/atomic_work_item_fence.html

> flags must be set to CLK_GLOBAL_MEM_FENCE, CLK_LOCAL_MEM_FENCE, CLK_IMAGE_MEM_FENCE 

Thinking about it, maybe it's better to use names consistent with our AMDGPUAS definitions, so image would just be "private" but local/global stay. What do you think?

Comment (arsenm) [reply clang/docs/LanguageExtensions.rst:4449]: Not sure we can get away without image

Comment (Pierre-vh) [reply clang/docs/LanguageExtensions.rst:4449]: This is just for internal use right now, so we can use anything we want. Device libs can then map the IMAGE flag to a "private" operand here.

File: llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp [hunk 14]
```diff
@@ -678,6 +680,54 @@ class SIMemoryLegalizer final : public MachineFunctionPass {
   bool runOnMachineFunction(MachineFunction &MF) override;
 };
 
+static std::array<std::pair<StringLiteral, SIAtomicAddrSpace>, 3> ASNames = {{
```
Comment (ssahasra) [inline llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp:683]: Use StringMap for this?

File: llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp [hunk 15]
```diff
@@ -678,6 +680,54 @@ class SIMemoryLegalizer final : public MachineFunctionPass {
...
+  auto MMRA = MMRAMetadata(MI.getMMRAMetadata());
+  if (!MMRA)
+    return Default;
+
+  SIAtomicAddrSpace Result = SIAtomicAddrSpace::NONE;
+  for (const auto &[Prefix, Suffix] : MMRA) {
+    if (Prefix != FenceASPrefix)
+      continue;
+
+    auto It = find_if(ASNames, [Suffix = Suffix](auto &Pair) {
```
Comment (ssahasra) [inline llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp:718]: I didn't understand the default assignment of Suffix. Shouldn't the lambda capture it simply because the outer symbol is used inside? But anyway, this use of find_if shouldn't be necessary with StringMap.

Comment (Pierre-vh) [reply llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp:718]: Lambdas can't capture structured bindings (`auto [a, b] = pair`) in C++17, I think it comes in C++20 or 23?

**Review by ssahasra — APPROVED** [review]
Looks good to me. But I have no opinion about that discussion with whether "image" should be available for explicit use!

**Comment by Pierre-vh** [comment]
@arsenm Should we use `image` or `private`?
We could allow both in the frontend, and only use `private` as the canonical MMRA.

**Comment by arsenm** [comment]
> @arsenm Should we use `image` or `private`? We could allow both in the frontend, and only use `private` as the canonical MMRA.

I don't understand why image would imply private. I would just keep at as private throughout

**Comment by Pierre-vh** [comment]
> > @arsenm Should we use `image` or `private`? We could allow both in the frontend, and only use `private` as the canonical MMRA.
> 
> I don't understand why image would imply private. I would just keep at as private throughout

I thought image memory = private. It's unclear to me, what AS does OpenCL IMAGE memory map to in our backend?
(But otherwise, yes, MMRA should just have the backend names, the mapping of the OpenCL IMAGE to a backend AS should be in the device-lib)

**Comment by arsenm** [comment]
> I thought image memory = private. It's unclear to me, what AS does OpenCL IMAGE memory map to in our backend? (But otherwise, yes, MMRA should just have the backend names, the mapping of the OpenCL IMAGE to a backend AS should be in the device-lib)

Images are global memory with magical addressing and value interpretation on load/store. There's nothing private about them

**Comment by Pierre-vh** [comment]
> > I thought image memory = private. It's unclear to me, what AS does OpenCL IMAGE memory map to in our backend? (But otherwise, yes, MMRA should just have the backend names, the mapping of the OpenCL IMAGE to a backend AS should be in the device-lib)
> 
> Images are global memory with magical addressing and value interpretation on load/store. There's nothing private about them

So in our case:

- OpenCL Image = Global
- OpenCL Local = LDS
- OpenCL global = Global 
?

Is that correct?

Then I guess the MMRA should just have "global" and "local" for now, we can always add more later if needed. What do you think?

**Comment by arsenm** [comment]
> Then I guess the MMRA should just have "global" and "local" for now, we can always add more later if needed. What do you think?

Yes, we don't have specific image counters. They are just vcmnt

**Comment by Pierre-vh** [comment]
> > Then I guess the MMRA should just have "global" and "local" for now, we can always add more later if needed. What do you think?
> 
> Yes, we don't have specific image counters. They are just vcmnt

Diff has been updated with those changes

File: llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp [hunk 16]
```diff
@@ -678,6 +680,49 @@ class SIMemoryLegalizer final : public MachineFunctionPass {
...
+static const StringMap<SIAtomicAddrSpace> ASNames = {{
+    {"global", SIAtomicAddrSpace::GLOBAL},
+    {"local", SIAtomicAddrSpace::LDS},
+}};
+
+void diagnoseUnknownMMRAASName(const MachineInstr &MI, StringRef AS) {
+  const MachineFunction *MF = MI.getMF();
+  const Function &Fn = MF->getFunction();
+  std::string Str;
+  raw_string_ostream OS(Str);
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIMemoryLegalizer.cpp:692]: SmallString + raw_svector_ostream?

**Review by arsenm — APPROVED** [review]
